cmd/dump/src/lib.rs

```diff
@@ -616,36 +693,154 @@ fn read_dump_header(
+    let op = hubris.get_idol_command("DumpAgent.read_dump")?;
+    let mut base = 0;
     let mut rval = vec![];
-    let mut done = false;
-    let mut ndx = 0;
 
-    while !done {
-        let header =
-            read_dump_at(hubris, core, context, funcs, ndx, |_| false)?;
+    //
+    // Iterate over dump area headers in chunksize units.  Once we've hit an
+    // area that hasn't been dumped to, or an invalid area (denoting that all
+    // areas are full), we're done...
+    //
+    loop {
+        let mut ops = vec![];
 
-        if header.next != 0 {
-            ndx += 1;
-        } else {
-            done = true;
+        for i in 0..chunksize {
+            let payload = op.payload(&[
+                ("index", idol::IdolArgument::Scalar((base + i) as u64)),
+                ("offset", idol::IdolArgument::Scalar(0u64)),
+            ])?;
+
+            context.idol_call_ops(funcs, &op, &payload, &mut ops)?;
+        }
+
+        ops.push(Op::Done);
+
+        let results = context.run(core, &ops, None)?;
+
+        for (i, r) in results.iter().enumerate() {
+            let area = base + i;
+
+            match r {
+                Ok(val) => {
+                    let header = parse_dump_header(base + i, val)?;
+
+                    if !raw && header.0.dumper == humpty::DUMPER_NONE {
+                        return Ok(rval);
+                    }
+
+                    rval.push(header);
+                }
+
+                Err(err) => {
+                    if op.strerror(*err) == "InvalidArea" {
+                        return Ok(rval);
+                    }
+
+                    bail!(
+                        "failed to header for area {area}: {}",
+                        op.strerror(*err)
+                    );
+                }
+            }
         }
 
-        rval.push(header);
+        base += chunksize;
     }
+}
+
+//
+// Because single task dumps can spread over adjacent areas, they can be a
+// little tedious to process; iterate over all dump areas and return a map of
+// area indices to a task/vector of headers tuple.
+//
+fn task_areas(
+    headers: &[(DumpAreaHeader, Option<DumpTask>)],
+) -> IndexMap<usize, (DumpTask, Vec<DumpAreaHeader>)> {
+    let mut rval = IndexMap::new();
 
-    Ok(rval)
+    if headers[0].1.is_none() {
+        return rval;
+    }
+
+    let mut current = 0;
+    let mut areas = vec![headers[0].0];
+
+    for (ndx, (header, task)) in headers.iter().enumerate().skip(1) {
+        match task {
+            None => {
+                //
+                // If the header indicates that it's been written, it's
+                // a continuation of our previous header -- otherwise, we're
+                // done processing entirely.
+                //
+                if header.dumper != humpty::DUMPER_NONE {
+                    areas.push(*header);
+                } else {
+                    break;
+                }
+            }
+            Some(_task) => {
```

Comment:
Add a comment that this marks the beginning of a new task dump, so we record the old one then start again.

Reply:
Done